codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/parse/EventStreamUnmarshallerGenerator.kt

```diff
@@ -0,0 +1,251 @@
+        "SmithyError" to RuntimeType("Error", CargoDependency.SmithyTypes(runtimeConfig), "smithy_types"),
+        "UnmarshallMessage" to RuntimeType("UnmarshallMessage", smithyEventStream, "smithy_eventstream::frame"),
+        "UnmarshalledMessage" to RuntimeType("UnmarshalledMessage", smithyEventStream, "smithy_eventstream::frame"),
+    )
+
+    fun render(): RuntimeType {
+        val unmarshallerType = unionShape.eventStreamUnmarshallerType()
+        return RuntimeType.forInlineFun("${unmarshallerType.name}::new", "event_stream_serde") { inlineWriter ->
+            inlineWriter.renderUnmarshaller(unmarshallerType, unionSymbol)
+        }
+    }
+
+    private fun RustWriter.renderUnmarshaller(unmarshallerType: RuntimeType, unionSymbol: Symbol) {
+        rust(
+            """
+            ##[non_exhaustive]
+            ##[derive(Debug)]
+            pub struct ${unmarshallerType.name};
+
+            impl ${unmarshallerType.name} {
+                pub fn new() -> Self {
+                    ${unmarshallerType.name}
+                }
+            }
+            """
+        )
+
+        rustBlockTemplate(
+            "impl #{UnmarshallMessage} for ${unmarshallerType.name}",
+            *codegenScope
+        ) {
+            rust("type Output = #T;", unionSymbol)
+            rust("type Error = #T;", operationErrorSymbol)
+
+            rustBlockTemplate(
+                """
+                fn unmarshall(
+                    &self,
+                    message: &#{Message}
+                ) -> std::result::Result<#{UnmarshalledMessage}<Self::Output, Self::Error>, #{Error}>
+                """,
+                *codegenScope
+            ) {
+                rustBlockTemplate(
+                    """
+                    let response_headers = #{Inlineables}::parse_response_headers(&message)?;
+                    match response_headers.message_type.as_str()
+                    """,
+                    *codegenScope
+                ) {
+                    rustBlock("\"event\" => ") {
+                        renderUnmarshallEvent()
+                    }
+                    rustBlock("\"exception\" => ") {
+                        renderUnmarshallError()
+                    }
+                    rustBlock("value => ") {
+                        rustTemplate(
+                            "return Err(#{Error}::Unmarshalling(format!(\"unrecognized :message-type: {}\", value)));",
+                            *codegenScope
+                        )
+                    }
+                }
+            }
+        }
+    }
+
+    private fun RustWriter.renderUnmarshallEvent() {
+        rustBlock("match response_headers.smithy_type.as_str()") {
+            for (member in unionShape.members()) {
+                val target = model.expectShape(member.target, StructureShape::class.java)
+                if (!target.hasTrait<ErrorTrait>()) {
```

Comment:
after reading this, I think we should use a model transformation to generate a synthetic union for the event stream operation that doesn't have the errors attached. It also removes the issue of the custom behavior in the union generator, removes the protocol dependence & allows us to add a custom trait to trigger the type change more easily.

Reply:
I like it!